File: tools/rlres/exporter/engine/anim/anim_serialize.py
# Copyright 2025 m4jr0. All Rights Reserved.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

from __future__ import annotations

from pathlib import Path
from typing import BinaryIO, Sequence

from rlres.data.engine.anim.anim_type import (
    MAX_ANIM_KEY_PARAM_COUNT,
    AnimKeyFrameModel,
    AnimResourceModel,
    AnimSampleModel,
    AnimSetResourceModel,
    AnimSolvedEntryModel,
)

from rlres.data.engine.param_set.param_set_serialize import write_param_set
from rlres.data.engine.resource.resource_type import ResourceTypeId
from rlres.exporter.engine.sprite.sprite_serialize import write_sprite

from rlres.utils.serialize_utils import (
    write_f32,
    write_file_header,
    write_u32,
    write_u64,
    write_u8,
    write_usize,
    write_vector_len,
)


def write_anim_sample(f: BinaryIO, s: AnimSampleModel) -> None:
    write_sprite(f, s.sprite)
    write_f32(f, s.duration)


def write_keyframes(f: BinaryIO, keys: Sequence[AnimKeyFrameModel]) -> None:
    write_vector_len(f, len(keys))

    for key in keys:
        write_u32(f, key.key_tag)
        write_u32(f, key.frame)
        write_param_set(f, key.params, MAX_ANIM_KEY_PARAM_COUNT)


def write_anim_resource(f: BinaryIO, a: AnimResourceModel) -> None:
    write_u32(f, a.flags)
    write_f32(f, a.duration)
    write_u64(f, a.key)

    write_vector_len(f, len(a.samples))

    for s in a.samples:
        write_anim_sample(f, s)

    write_vector_len(f, len(a.acc_durations))

    for d in a.acc_durations:
        write_f32(f, d)

    write_keyframes(f, a.keys)


def write_anim_solved(f: BinaryIO, solver: list[AnimSolvedEntryModel]) -> None:
    write_vector_len(f, len(solver))

    for entry in solver:
        write_u64(f, int(entry.from_key))
        write_u64(f, int(entry.to_key))
        write_u8(f, int(entry.flip_policy))


def write_anim_set_resource(f: BinaryIO, aset: AnimSetResourceModel) -> None:
    write_u32(f, aset.rid)
    write_u32(f, aset.tex)
    write_u32(f, aset.collider_profile)
    write_usize(f, aset.default_anim_idx)

    write_vector_len(f, len(aset.anims))

    for a in aset.anims:
        write_anim_resource(f, a)

    write_anim_solved(f, aset.solved)


def write_anim_set_file(path: Path, aset: AnimSetResourceModel) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)

    with path.open("wb") as f:
        write_file_header(f, ResourceTypeId.ANIM_SET, AnimSetResourceModel.VERSION)
        write_anim_set_resource(f, aset)

File: src/engine/anim/anim_serialize.cc
// Copyright 2025 m4jr0. All Rights Reserved.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

// Precompiled. ////////////////////////////////////////////////////////////////
#include "precompiled.h"
////////////////////////////////////////////////////////////////////////////////

// Header. /////////////////////////////////////////////////////////////////////
#include "anim_serialize.h"
////////////////////////////////////////////////////////////////////////////////

#include "engine/anim/anim_resource.h"
#include "engine/anim/anim_type.h"
#include "engine/resource/resource_file_serialize.h"
#include "engine/sprite/sprite_serialize.h"

namespace rl {
void writeAnimSample(std::ostream& os, const AnimSampleResource& s) {
  writeSprite(os, s.sprite);
  writePod(os, s.duration);
}

void readAnimSample(std::istream& is, AnimSampleResource& s) {
  readSprite(is, s.sprite);
  readPod(is, s.duration);
}

void writeAnimSampleVector(std::ostream& os,
                           const std::vector<AnimSampleResource>& v) {
  writeVectorWith(os, v, [](std::ostream& os2, const AnimSampleResource& s) {
    writeAnimSample(os2, s);
  });
}

void readAnimSampleVector(std::istream& is,
                          std::vector<AnimSampleResource>& v) {
  readVectorWith(is, v, [](std::istream& is2, AnimSampleResource& s) {
    readAnimSample(is2, s);
  });
}

void writeParamSet(std::ostream& os, const AnimParamSet& ps) {
  writePod(os, ps);
}

void readParamSet(std::istream& is, AnimParamSet& ps) { readPod(is, ps); }

void writeAnimKeyFrame(std::ostream& os, const AnimKeyFrame& kf) {
  writePod(os, kf.keyTag);
  writePod(os, kf.frame);

  const AnimParamSet& base = static_cast<const AnimParamSet&>(kf);
  writeParamSet(os, base);
}

void readAnimKeyFrame(std::istream& is, AnimKeyFrame& kf) {
  readPod(is, kf.keyTag);
  readPod(is, kf.frame);

  AnimParamSet& base = static_cast<AnimParamSet&>(kf);
  readParamSet(is, base);
}

void writeAnimKeyFrameVector(std::ostream& os,
                             const std::vector<AnimKeyFrame>& v) {
  writeVectorWith(os, v, [](std::ostream& os2, const AnimKeyFrame& kf) {
    writeAnimKeyFrame(os2, kf);
  });
}

void readAnimKeyFrameVector(std::istream& is, std::vector<AnimKeyFrame>& v) {
  readVectorWith(is, v, [](std::istream& is2, AnimKeyFrame& kf) {
    readAnimKeyFrame(is2, kf);
  });
}

void writeAnimResource(std::ostream& os, const AnimResource& a) {
  writePod(os, a.flags);
  writePod(os, a.duration);
  writePod(os, a.key);

  writeAnimSampleVector(os, a.samples);
  writeVector(os, a.accDurations);
  writeAnimKeyFrameVector(os, a.keys);
}

void readAnimResource(std::istream& is, AnimResource& a) {
  readPod(is, a.flags);
  readPod(is, a.duration);
  readPod(is, a.key);

  readAnimSampleVector(is, a.samples);
  readVector(is, a.accDurations);
  readAnimKeyFrameVector(is, a.keys);
}

void writeAnimSolvedEntry(std::ostream& os, const AnimSolvedEntry& e) {
  writePod(os, e.fromKey);
  writePod(os, e.toKey);
  writePod(os, e.flipPolicy);
}

void readAnimSolvedEntry(std::istream& is, AnimSolvedEntry& e) {
  readPod(is, e.fromKey);
  readPod(is, e.toKey);
  readPod(is, e.flipPolicy);
}

void writeAnimSolvedEntryVector(std::ostream& os,
                                const std::vector<AnimSolvedEntry>& v) {
  writeVectorWith(os, v, [](std::ostream& os2, const AnimSolvedEntry& e2) {
    writeAnimSolvedEntry(os2, e2);
  });
}

void readAnimSolvedEntryVector(std::istream& is,
                               std::vector<AnimSolvedEntry>& v) {
  readVectorWith(is, v, [](std::istream& is2, AnimSolvedEntry& e2) {
    readAnimSolvedEntry(is2, e2);
  });
}

void writeAnimSetResource(std::ostream& os, const AnimSetResource& set) {
  writePod(os, set.id);
  writePod(os, set.tex);
  writePod(os, set.colliderProfile.id);
  writePod(os, set.defaultAnimIdx);
  writeAnimSolvedEntryVector(os, set.solved);

  writeVectorWith(os, set.anims, [](std::ostream& os2, const AnimResource& a) {
    writeAnimResource(os2, a);
  });
}

void readAnimSetResource(std::istream& is, AnimSetResource& set) {
  readPod(is, set.id);
  readPod(is, set.tex);
  readPod(is, set.colliderProfile.id);
  readPod(is, set.defaultAnimIdx);

  readVectorWith(is, set.anims, [](std::istream& is2, AnimResource& a) {
    readAnimResource(is2, a);
  });

  readAnimSolvedEntryVector(is, set.solved);

  set.keyToIdx.clear();
  set.keyToIdx.reserve(set.anims.size());

  for (usize i = 0; i < set.anims.size(); ++i) {
    const AnimResource& a = set.anims[i];
    set.keyToIdx[a.key] = i;
  }
}

void writeAnimSetFile(std::ostream& os, const AnimSetResource& set) {
  writeResourceFileHeader(os, kResourceTypeAnimSet, AnimSetResource::kVersion);
  writeAnimSetResource(os, set);
}

bool readAnimSetFile(std::istream& is, AnimSetResource& set, u32& outVersion) {
  if (!readAndCheckResourceFileHeader(is, kResourceTypeAnimSet, outVersion)) {
    return false;
  }

  readAnimSetResource(is, set);
  return AnimSetResource::kVersion == outVersion;
}
}  // namespace rl